### PR TITLE
frontend-*-api: reimplement routing system from core-*-api

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Work towards #19545

Looked at this for a long while and figured that the best path forward here is to introduce a new set of route ref types in the new packages. This lets us hide a few fields and move to the new and less arcane way of defining opaque types. The new route refs don't require IDs, as they will get their IDs assigned to them by the app according to how they are exposed via their plugin. In cases where route refs have not yet been assigned an ID the declaration location in code will instead be used to describe the the route ref in order to aid debugging.

This PR removes the requirement for route refs to have the same ID as their extension in the new system.

Several types and fields are also deprecated in the `core-plugin-api` types that no longer exist in the `frontend-plugin-api` types. The idea is that as long as no deprecated types and fields are used, `frontend-plugin-api` will be a drop-in replacement.

I felt that the `AnyParams` types also had too generic of a name. It really only makes sense to use as a type argument for the route ref types, which is also what it is intended for. Initially I called it `AnyRouteParams` but then figured that `AnyRouteRefParams` was even better, since that's exactly what it's for. This is also changed in the existing system and the old type is deprecated.

There's a `convertLegacyRouteRef` helper from `@backstage/core-plugin-api/alpha` that lets us use old route refs in the new system for now. The idea is to move this over to a new compatibility package, but also to completely remove the need for it by making the legacy route refs implement the new type once the API is stable enough.

The plan is the following:
- New plugins are built completely on the new system and types and there is no backwards compatibility code in the new system.
- Existing plugins can pass their existing refs to the new system, initially through `convertLegacyRouteRef`, but in the future it'll work even without that.
- Once an existing plugin is fully migrated to the new system it will no longer be possible to use it in the old one, unless we add legacy compatibility helpers.
- Existing plugins that are still built on the old system and still depend on an older version of `core-plugin-api` can be installed in new apps through our compatibility layer, which also provides a context with conversion from legacy to the new route refs.

As a side-note I'm planning a refactor that removes the routing context in the new system and instead provide a Utility API that handles routing, which useRouteRef will use instead. Remains to be seen whether we need to keep the context around then or not, or it can be completely moved over to the legacy compatibility APIs as along as we change `useRouteRef` in `core-plugin-api` too.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/frontend-app-api/src/routing/resolveRouteBindings.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    bindRoutes({ bind });
+  }
+
+  const bindingsConfig = config.getOptionalConfig('app.routes.bindings');
+  if (!bindingsConfig) {
+    return result;
+  }
+
+  const bindings = bindingsConfig.get();
+  if (bindings === null || typeof bindings !== 'object') {
```
Comment (freben) [inline packages/frontend-app-api/src/routing/resolveRouteBindings.ts:113]: `getOptionalConfig` already ensured this, should be ok to put a type arg on `get` and be done

**Review by freben — COMMENTED** [review]
amaze

File: packages/frontend-app-api/src/routing/collectRouteIds.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  BackstagePlugin,
+  ExtensionOverrides,
+  RouteRef,
+  SubRouteRef,
+  ExternalRouteRef,
+} from '@backstage/frontend-plugin-api';
+// eslint-disable-next-line @backstage/no-relative-monorepo-imports
+import { toInternalRouteRef } from '../../../frontend-plugin-api/src/routing/RouteRef';
+// eslint-disable-next-line @backstage/no-relative-monorepo-imports
+import { toInternalExternalRouteRef } from '../../../frontend-plugin-api/src/routing/ExternalRouteRef';
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/frontend-app-api/src/routing/collectRouteIds.ts:27]: `InternalValueDestructorExternalInterfaceConstructorValueRef`

File: packages/frontend-app-api/src/routing/collectRouteIds.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  const routesById = new Map<string, RouteRef | SubRouteRef>();
+  const externalRoutesById = new Map<string, ExternalRouteRef>();
+
+  for (const feature of features) {
+    if (feature.$$type !== '@backstage/BackstagePlugin') {
+      continue;
+    }
+
+    for (const [name, ref] of Object.entries(feature.routes)) {
+      const refId = `plugin.${feature.id}.routes.${name}`;
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/frontend-app-api/src/routing/collectRouteIds.ts:48]: Yes we should x)

File: plugins/tech-radar/alpha-api-report.md [hunk 4]
```diff
@@ -3,13 +3,18 @@
...
+import { AnyExternalRoutes } from '@backstage/frontend-plugin-api';
 import { BackstagePlugin } from '@backstage/frontend-plugin-api';
 import { Extension } from '@backstage/frontend-plugin-api';
+import { RouteRef } from '@backstage/frontend-plugin-api';
 
 // @alpha (undocumented)
-const _default: BackstagePlugin<AnyRoutes, AnyExternalRoutes>;
+const _default: BackstagePlugin<
+  {
+    root: RouteRef<undefined>;
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/tech-radar/alpha-api-report.md:14]: Yep I think we should cover this as we look over the extension naming

File: plugins/graphiql/alpha-api-report.md [hunk 5]
```diff
@@ -21,7 +21,12 @@ export function createEndpointExtension<TConfig extends {}>(options: {
 }): Extension<TConfig>;
 
 // @alpha (undocumented)
-const _default: BackstagePlugin<AnyRoutes, AnyExternalRoutes>;
+const _default: BackstagePlugin<
+  {
+    root: RouteRef<undefined>;
+  },
+  AnyExternalRoutes
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/graphiql/alpha-api-report.md]: Yep! done, figured `{}` is the most correct

File: packages/frontend-app-api/src/routing/resolveRouteBindings.ts (hunk 1 above)
Comment (Rugvip) [reply packages/frontend-app-api/src/routing/resolveRouteBindings.ts]: Ah yep, realized in tests but forgot to clean this bit up x)

**Review by freben — APPROVED** [review]
aight!

**Pull request merged**